datafusion/core/tests/sqllogictests/test_files/groupby.slt

```diff
@@ -2384,3 +2384,197 @@ SELECT s.country, ARRAY_AGG(s.amount ORDER BY s.country DESC, s.amount DESC) AS
 FRA [200.0, 50.0] 250
 GRC [80.0, 30.0] 110
 TUR [100.0, 75.0] 175
+
+# test_reverse_aggregate_expr
+# Some of the Aggregators can be reversed, by this way we can still run aggregators
+# that have contradictory requirements at first glance.
+query TT
+EXPLAIN SELECT country, ARRAY_AGG(amount ORDER BY amount DESC) AS amounts,
+  FIRST_VALUE(amount ORDER BY amount ASC) AS fv1,
+  LAST_VALUE(amount ORDER BY amount DESC) AS fv2
+  FROM sales_global
+  GROUP BY country
+----
+logical_plan
+Projection: sales_global.country, ARRAY_AGG(sales_global.amount) ORDER BY [sales_global.amount DESC NULLS FIRST] AS amounts, FIRST_VALUE(sales_global.amount) ORDER BY [sales_global.amount ASC NULLS LAST] AS fv1, LAST_VALUE(sales_global.amount) ORDER BY [sales_global.amount DESC NULLS FIRST] AS fv2
+--Aggregate: groupBy=[[sales_global.country]], aggr=[[ARRAY_AGG(sales_global.amount) ORDER BY [sales_global.amount DESC NULLS FIRST], FIRST_VALUE(sales_global.amount) ORDER BY [sales_global.amount ASC NULLS LAST], LAST_VALUE(sales_global.amount) ORDER BY [sales_global.amount DESC NULLS FIRST]]]
+----TableScan: sales_global projection=[country, amount]
+physical_plan
+ProjectionExec: expr=[country@0 as country, ARRAY_AGG(sales_global.amount) ORDER BY [sales_global.amount DESC NULLS FIRST]@1 as amounts, FIRST_VALUE(sales_global.amount) ORDER BY [sales_global.amount ASC NULLS LAST]@2 as fv1, LAST_VALUE(sales_global.amount) ORDER BY [sales_global.amount DESC NULLS FIRST]@3 as fv2]
+--AggregateExec: mode=Single, gby=[country@0 as country], aggr=[ARRAY_AGG(sales_global.amount), FIRST_VALUE(sales_global.amount), LAST_VALUE(sales_global.amount)]
```

Comment:
I don't understand this plan -- if the sort is by `amount1 DESC` then isn't 

I am expecting that these are the same:

```sql
  FIRST_VALUE(amount ORDER BY amount ASC) AS fv1,
  LAST_VALUE(amount ORDER BY amount DESC) AS fv2
```

So when the orderby is rewritten to be `ORDER BY amount1 DESC` I expect to see that 

```sql
  LAST_VALUE(amount ORDER BY amount DESC) AS fv1, -- this got switched
  LAST_VALUE(amount ORDER BY amount DESC) AS fv2
```

However, the query seems to get the right answer, so I must be missing something

Reply:
> > in terms of behavior. Hence by converting query above to its equivalent form below
> 
> ```sql
> SELECT country, LAST_VALUE(amount ORDER BY amount DESC) AS fv1,
>   LAST_VALUE(amount ORDER BY amount DESC) AS fv2,
>   ARRAY_AGG(amount ORDER BY amount ASC) AS amounts
>   FROM sales_global
>   GROUP BY country
> ```
> 
> Did you mean like this (change `LAST_VALUE` to `FIRST_VALUE` and switch the order):
> 
> ```sql
> SELECT country, LAST_VALUE(amount ORDER BY amount DESC) AS fv1,
>   FIRST_VALUE(amount ORDER BY amount ASC) AS fv2,
>   ARRAY_AGG(amount ORDER BY amount ASC) AS amounts
>   FROM sales_global
>   GROUP BY country
> ```

Sorry, I have updated the example in the PR body (changed requirement of `ARRAY_AGG` from `amount ASC` to `amount DESC`). In the equivalent form all requirements, of aggregate functions are same. Hence we have resolved conflicting requirements by conversion. The conversion takes place between `FIRST_VALUE` and `LAST_VALUE` and ordering requirement is switched as you say. For instance `FIRST_VALUE(a ORDER BY b DESC)` is equivalent to `LAST_VALUE(a ORDER BY b ASC)`

Reply:
`AggregateExpr` has a `fn name(&self) -> &str` method. This method returns string representation of the expression. This is given during initialization from outside(It is not calculated according to function and arguments). During reversal we use the name as is. However, it is misleading, as you observed (even though behind the scene appropriate version is working). However, I couldn't find an easy fix for this problem

Reply:
🤔 I wonder if during rewrite we could rewrite the value of `name()` somehow

Reply:
We can replace `FIRST_VALUE` with `LAST_VALUE` vice versa inside the string. This is not a nice solution, but this would solve our problem, what do you think about it?

Reply:
I think, we can rewrite `name` similar to the version that is given during initialization. This way we may not overfit on string convention. I will experiment with rewriting the value of `name()`, and let you know about my observations.

Reply:
@alamb I have changed the `name` implementation. It is now calculated inside the aggregator, it is not given from outside during initialization. The display of the AggregateExec changed from 
`--AggregateExec: mode=Single, gby=[country@0 as country], aggr=[ARRAY_AGG(sales_global.amount), FIRST_VALUE(sales_global.amount), LAST_VALUE(sales_global.amount)]` to 
`--AggregateExec: mode=Single, gby=[country@0 as country], aggr=[ARRAY_AGG(sales_global.amount), LAST_VALUE(amount@1), LAST_VALUE(amount@1)]`

Since argument expressions are `Arc<dyn PhysicalExpr>`, output display of the argument is changed from `sales_global.amount` to `amount@1` for `FIRST_VALUE` AND `LAST_VALUE`. 

This change, fortunately lead me to recognize a bug during order-sensitive aggregation when partition number is greater than 1.
In multi partition cases, it is not guaranteed that `Final` or `FinalPartitioned` stages in the `AggregateExec` to receive data in-order. 
Hence I constrained order-sensitive aggregation to single partition cases. 
I will add support for order-sensitive aggregation at multiple partitions, in following PRs.

Reply:
So we have two options:
1. Simple string replacement of `FIRST_VALUE` to `LAST_VALUE` (and vice versa) and order direction when reversal happens.
2. Generating name with the new mechanism @mustafasrepo experimented with.

The advantage of (1) is that it preserves a uniform display format. Its disadvantage is that this is a specialized solution to `FIRST_VALUE`/`LAST_VALUE` and may not generalize to other functions in the future.

The advantage of (2) is that is generalizes, but it has the drawback that we end up with an intermixed format; i.e. we have entries like `ARRAY_AGG(sales_global.amount)` along with entries like `LAST_VALUE(amount@1)`.

Let's choose one for now and create an issue to fix this in a more general fashion in the future. I do not have a preference at this point -- @alamb, which one do you prefer?

Reply:
I don't have any preference -- they both sound fine to me